### ISSUE TITLE
Add selector-no-empty rule

### ISSUE THREAD

**Issue by silverwind** [stylelint/stylelint#1777]
Are syntax error in scope of this linter? I'd imagine a rule that catches trailing commas in selector lists like

```css
.a, .b, {
  color: red;
}
```
would be quite useful. Here's a [recent case](https://github.com/StylishThemes/GitHub-Dark/commit/e5bf405f2af578c5feed849893c56dc575deb37d).

**Comment by m-allanson** [comment]
I think this meets [the criteria for new rules](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#let-everyone-know) and sounds like a useful addition. Let's see if stylelint/core have any thoughts on this.

**Comment by evilebottnawi** [comment]
@silverwind it is valid css?

**Comment by evilebottnawi** [comment]
Not valid syntax, mayba create `selector-no-invalid` and check on permitted characters in selector and commas, ignoring nonstandard css

**Comment by jeddy3** [comment]
> maybe create selector-no-invalid and check on permitted characters in selector and commas, ignoring nonstandard css

SGTM. We've used `no-invalid` else where for syntax errors *that PostCSS can parse*.

**Comment by silverwind** [comment]
`selector-no-invalid` sounds good. It could also catch empty selector list elements anywhere:

```css
, .a, .b { color: red; } /* empty element at start */
.a, .b,  { color: red; } /* empty element at end */
.a, , .b { color: red; } /* empty element in the middle */
```
(Those are all invalid)

**Comment by evilebottnawi** [comment]
@silverwind Also need check on all invalid characters, sgtm

**Comment by jeddy3** [comment]
Any objections to labelling this up as "help wanted"?

**Comment by evilebottnawi** [comment]
sgtm

**Comment by jeddy3** [comment]
@silverwind Would you like to contribute this rule?

**Comment by silverwind** [comment]
I'll try.

**Comment by davidtheclark** [comment]
`selector-no-invalid` seems like a super general name. I think we might want to limit the scope of the rule by saying precisely what invalid pattern is being caught — otherwise it's kind of implied that the rule should catch any form of invalidity.

**Comment by jeddy3** [comment]
> I think we might want to limit the scope of the rule by saying precisely what invalid pattern is being caught 

Agreed. I had the same thought myself and I was going to come back to this today :)

`selector-no-invalid-commas` or `selector-no-extra-commas`?

**Comment by silverwind** [comment]
I don't think "commas" should really be in the name. Ideally, I'd also like it to catch a empty selector like

````css
{ color: red}
````
Maybe `selector-no-empty`?

**Comment by davidtheclark** [comment]
I don't think there is enough clarity in what exactly this rule is intended to enforce. We need to articulate that before anybody starts trying to build it. What are the failing cases, what are the passing cases, and what is the logic that distinguishes between them?

**Comment by jeddy3** [comment]
> We need to articulate that before anybody starts trying to build it

Agreed.

I think @silverwind's suggestion of `selector-no-empty` is perhaps on the right path. 

- Name: `selector-no-empty`
- Description: "Disallow empty selectors"
- Extended description/logic: "Empty selectors (by themselves or within a selector list) invalidate a rule. This rule splits the selector list on `,` and anything part of zero length or contains only whitespace is consider empty. Everything else isn't."
- Primary options: `true`
- Secondary options: none
- Message: `Unexpected empty selector`
- Section: "Selector"

Failing patterns:

```css
{ color: red} /* entirely empty selector */
, .a, .b { color: red; } /* empty selector at the start */
.a, .b,  { color: red; } /* empty selector at the end */
.a, , .b { color: red; } /* empty selector in the middle */
```

Passing patterns:

```css
.a { color: red; }
.a, .b { color: red }
```

Does that feel more defined and of a known scope?

**Comment by silverwind** [comment]
That looks great, @jeddy3, just what I imagined. The intention is to catch the common error of empty selectors (which are often introduced by extra commas).

A possible future `selector-no-invalid` could supersede and replace `selector-no-empty` eventually, but I estimate that determining if a selector is valid is a **major** undertaking.

**Comment by davidtheclark** [comment]
Great, thanks! I'll setup the labels.

**Comment by silverwind** [comment]
I don't think I'll have time to work on this rule in the coming weeks, so if anyone wants to take it on, please go ahead.

**Comment by evilebottnawi** [comment]
@silverwind okey

**Comment by gaidarenko** [comment]
Please label this issue as `wip`

**Comment by evilebottnawi** [comment]
@gaidarenko :+1:
